Ignore ControlPersist broken pipe errors. Fixes #16731

kustodian opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### ANSIBLE VERSION

```
ansible 2.2.0
```
##### SUMMARY

Automatically retires an ssh connection if it broke because of `mux_client_hello_exchange: write packet: Broken pipe` which happens when Ansible attempts to reuse an SSH ControlMaster socket just when the ControlPersist time out expires. More about the issue this pull request fixes in #16731.

jctanner commented:
@kustodian hey, thanks for writing this up. I would like for it to also provide the full errors for other problems too.

kustodian commented:
@jctanner this won't display actual SSH errors which you want to add into
Ansible, I left that part to you, since you said you will fix that :-)

This only fixes the bug with Ansible disconnecting because of
ControlPersist timeout expiring when Ansible is just about to connect,
which causes random host disconnects and is explained in detail in #16731.

abadger commented:
@bcoca has given this a +1 on slack.

With kustodian's explanation: https://github.com/ansible/ansible/issues/16731#issuecomment-291443094 I think this looks okay as well.

Waiting for @nitzmahone to have a chance to look at this and then can decide whether it's ready for merge.

abadger commented:
nitzamahone also gives +1 with the explanation here: https://github.com/ansible/ansible/issues/16731#issuecomment-291443094

nitzmahone commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -550,8 +550,14 @@ def _run(self, cmd, in_data, sudoable=True):
+        # we raise an internal error so that we can retry a connection.
+        controlpersist_broken_pipe = 'mux_client_hello_exchange: write packet: Broken pipe' in stderr
+        if p.returncode == 255 and controlpersist_broken_pipe:
+            raise AnsibleInternalError('SSH Error: data could not be sent because of ControlPersist broken pipe.')
```
Since this is a generic exception, I think we need a specific error code or something to ensure that if someone else tosses this down the line for another reason, that the associated retry doesn't get triggered.

nitzmahone commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -596,10 +602,15 @@ def exec_command(self, *args, **kwargs):
+                    # 255 = failure from the ssh command itself
+                except (AnsibleInternalError) as e:
+                    # Retry one more time because of the ControlPersist broken pipe
+                    display.vvv(u"RETRYING BECAUSE OF CONTROLPERSIST BROKEN PIPE")
```
Ensure that it's the exact right exception (by some mechanism, eg, error code as mentioned above) before we trigger the retry.

nitzmahone commented:
I'm OK with the overall approach, just want to make sure we don't trigger this inadvertently in the future if someone else throws `AnsibleInternalError` for some other reason. Also perhaps some more caution tape around the actual retry, saying that it's only safe to use in this exact scenario.

kustodian committed:
Ignores ControlPersist broken pipe errors. Fixes #16731

Ansible will now automatically retry a connection if SSH returns an error:

mux_client_hello_exchange: write packet: Broken pipe

This is probably a bug in SSH, but because it's safe to retry this
connection there is no need for Ansible to fail because of it.

kustodian commented:
Because in the mean time `ssh.py` was changed a lot, it was very hard for me to do a proper merge. It was a lot easier to apply my changes to the latest version of the file.

@nitzmahone I created a separate class for this exception, so it this type of exception won't get raised by a mistake by something else, so I guess this should solve your concern.

kustodian commented:
@nitzmahone I would really appreciate if you could help me with the two unit tests which are failing, I don't even know where to start.

abadger commented:
Merged to devel.
